### DISCUSSION TITLE
Can we integrate Kernal 6.1.0 and Vad5050 0.21.0 into one project

### DISCUSSION THREAD

**Discussion by lusunday** [openTCS/opentcs#173] [Q&A] [answered]
Because I need to modify the code of vda5050 and the kernel, I chose to compile the code of vda5050 and add it to the kernel project. Since I am not using a Fat JAR or the like, I manually introduced the "everit json schema" and "mqttv3" required by vda5050. The rest depends on it pointing directly to my current kernel project. Currently, the project can start normally, but I am not sure if there will be any problems in the future because vda5050's 0.21.0 version should depend on kernel 6.0.0. If there are no problems, should I directly import vda5050's code as a module into the kernel instead of introducing a jar

**Comment by swltr** [comment]
@lusunday It is not quite clear for me how exactly you have set up your project, but it sounds like you had to integrate the JSON Schema and MQTT dependencies manually. The [example integration project](https://github.com/openTCS/opentcs-integration-example) is supposed to be a reference for how to set up your own project.

**Reply by lusunday** [reply]
I downloaded the source code of vda5050 version 0.21.0, built it, and imported the jar into the kernel project. After using gradle build, the jar does not include the original dependencies, so I had to manually add the dependencies originally required by vda5050 in the kernel project to start the kernel project. I only import "everit json schema" and "mqttv3", and successfully started the kernel project. However, the opentcs-api-injection and opentcs-common, which were originally dependencies of vda5050, were effectively forced from version 6.0.0 to 6.1.0 because my kernel project is version 6.1.


This is the kernel's build.gradle, nothing else has changed
dependencies {
  api project(':openTCS-API-Injection')
  api project(':openTCS-Common')
  api project(':openTCS-CommAdapter-Loopback')
  api project(':openTCS-PeripheralCommAdapter-Loopback')
  api project(':openTCS-Strategies-Default')
  api project(':openTCS-Impl-Configuration-gestalt')
  api project(':openTCS-Kernel-Extension-HTTP-Services')
  api project(':openTCS-Kernel-Extension-RMI-Services')

  implementation group: 'de.huxhorn.sulky', name: 'de.huxhorn.sulky.ulid', version: '8.3.0'
  
  runtimeOnly group: 'org.slf4j', name: 'slf4j-jdk14', version: '2.0.13'
  
  api group: 'org.eclipse.paho', name: 'org.eclipse.paho.client.mqttv3', version: '1.2.5'
  api group: 'com.github.erosb', name: 'everit-json-schema', version: '1.14.3'
  implementation files("${rootDir}/libs/opentcs-commadapter-vda5050-0.21.0-SNAPSHOT.jar")
}

**Reply by swltr** [reply]
Yes, when manually downloading the JAR and using it as a dependency, you have to manage transitive dependencies yourself.

As mentioned, the example integration project shows the intended way to set up an openTCS project. My recommendation is to base your project on that.

**Reply by lusunday** [reply]
After using the example integration project to build the project, there is no Java code in the main of the kernel, so I am unable to modify the source code.

Additionally, I found that adding the following line to the `build.gradle` file of the kernel:
implementation 'org.opentcs.commadapter.vehicle.vda5050:opentcs-commadapter-vda5050:0.21.0'
causes the kernel's own `opentcs-api-injection` and `opentcs-common` dependencies to point to the remote 6.0.0 versions.


I am out of options and feel that I can only wait for the vda5050 baseline to be updated to 6.1.0, or downgrade my kernel version to 6.0.0 in order to modify the source code.

**Reply by swltr** [reply, marked as answer]
Right - using the example integration project is recommended exactly because it separates your code from that of the dependencies, and because the idea is that you work with the openTCS API to extend and customize openTCS. That way, you only have to update your dependencies to update e.g. to a new version of openTCS or the VDA5050 driver. If you instead import the openTCS and driver source tree into your own project and modify the sources, you're basically on your own.
